internal/controllers/machineset/machineset_controller.go

```diff
@@ -218,6 +220,18 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		})
 	}
 
+	// If the MachineSet is topology-managed ensure the MachineSetTopologyFinalizer is reconciled.
+	if feature.Gates.Enabled(feature.ClusterTopology) {
+		if _, ok := machineSet.Labels[clusterv1.ClusterTopologyOwnedLabel]; ok {
+			// - If the MachineSet is not being deleted.
+			if machineSet.DeletionTimestamp.IsZero() ||
+				// - If the MachineSet is being deleted but currently still has the Finalizer.
+				(!machineSet.DeletionTimestamp.IsZero() &&
+					controllerutil.ContainsFinalizer(machineSet, clusterv1.MachineSetTopologyFinalizer)) {
+				controllerutil.AddFinalizer(machineSet, clusterv1.MachineSetTopologyFinalizer)
+			}
```

Comment:
Given that we are not using SSA, I will simplify this and just add the finalizer back if the machine is not being deleted

Reply:
Sounds good to me

Reply:
Given that this finalizer is actually owned by the MS **topology** controller. What about setting it in the MS topology controller?

Reply:
+1 to moving to the topology controller, it'd be consistent with the other PR as well #7531